### ISSUE TITLE
Support custom gradient reduction algorithm in DDP

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#37359]
Currently, DDP always computes the average all gradients across all processes, which may not always be mathematically equivalent to local training. For example, if the loss function is MSE in sum mode, the gradient reduction should use sum instead of mean. If the different processes use different batch size, DDP might need to communicate the batch size as well to compute the weighted mean. 

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar

**Comment by rohan-varma** [comment]
For different processes using different batch sizes, won't we need to resolve https://github.com/pytorch/pytorch/issues/33148 before doing this?

**Comment by zhaojuanmao** [comment]
is it not a bug if loss function is MSE and DDP is using mean?

**Comment by mrshenli** [comment]
> For different processes using different batch sizes, will we need to resolve #33148 to support uneven batch sizes before doing this?

Sorry that I didn't state that clearly. No, we don't need to resolve #33148 first. By batch size I mean the number of samples in one batch. The issue tracked in #33148 is for different number of batches/iterations. More specially, if we have 2 DDP processes, where one processing 1 batch and another processing 2 batch, it would hang. But if both of them process 1 batch and one process's batch contains 10 sample and another contains 20 sample, it won't hang. However, the gradient might need to be averaged in a weighted manner depending on the loss function, sth like `(10 * grad1 + 20 * grad2) / 30`.

**Comment by mrshenli** [comment]
> is it not a bug if loss function is MSE and DDP is using mean?

If we consider mathematical in-equivalence as a bug, then yes, this is bug. But this is not easy to come up with a general fix that works for all loss functions. For MSE, it's easier, we probably can support sum/mean/weighted-sum mode in DDP. But for customized loss functions, we would never now what would be the correct way to do reduction to maintain mathematical equivalence. So might need to make this customizable.

**Comment by mrshenli** [comment]
addressed by #39272

**Comment by ProHuper** [comment]
@mrshenli Hi, did you solve the problem? I met the same problem here, can you give me an example of gradient reduction in sum mode? thx.

**Comment by hhc1997** [comment]
@ProHuper  Hi, I have the same problem when using sum loss. Did you solve this?
